### ISSUE TITLE
ListView Builder scrolling is slow

### ISSUE THREAD

**Issue by kieferlam** [flutter/flutter#22314]
The ListView Builder is slow compared to native apps. I've attached a video demo-ing my experience. And my complete code (which is basically just the example on the `Working with long lists guide`.

The gif has the flutter listview and also shows the scrolling of Reddit Is Fun app as well just for a reference.

[gif](https://giant.gfycat.com/DearestAccurateCanine.webm)

main.dart

```dart
import 'package:flutter/material.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Flutter Demo',
      home: Scaffold(
        body: ListView.builder(
            key: Key("lvbuildertest"),
            itemCount: 200,
            itemBuilder: (context, index) {
              return ListTile(title: Text("index: $index"));
            }),
      ),
    );
  }
}

```

```
[√] Flutter (Channel master, v0.9.4-pre.31, on Microsoft Windows [Version 10.0.17134.285], locale en-GB)
    • Flutter version 0.9.4-pre.31 at F:\androidsdk\flutter
    • Framework revision 52697ef962 (11 hours ago), 2018-09-26 07:57:05 +0200
    • Engine revision 38a646e14c
    • Dart version 2.1.0-dev.5.0.flutter-4cf2d3990b

[√] Android toolchain - develop for Android devices (Android SDK 28.0.2)
    • Android SDK at F:\androidsdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.2
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b02)
    • All Android licenses accepted.

[√] Android Studio (version 3.1)
    • Android Studio at C:\Program Files\Android\Android Studio
    X Flutter plugin not installed; this adds Flutter specific functionality.
    X Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b02)

[!] IntelliJ IDEA Ultimate Edition (version 2018.1)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA 2018.1.5
    X Flutter plugin not installed; this adds Flutter specific functionality.
    X Dart plugin not installed; this adds Dart specific functionality.
    • For information about installing plugins, see
      https://flutter.io/intellij-setup/#installing-the-plugins

[√] VS Code (version 1.27.2)
    • VS Code at C:\Users\username


\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 2.18.0

[√] Connected devices (1 available)
    • SM N9005 • 192.168.0.3:5555 • android-arm • Android 8.1.0 (API 27)

! Doctor found issues in 1 category.
```

**Comment by zoechi** [comment]
#20857 might be related

**Comment by iamarnas** [comment]
@kieferlam Have you tried to add "physics: const AlwaysScrollableScrollPhysics ()," it will make smother scrolling.

**Comment by kieferlam** [comment]
@iamarnas That has definitely improved scrolling performance! Thanks.

**Comment by bobmoff** [comment]
I am curious to why AlwaysScrollableScrollPhysics would improve the performance of scrolling ? Isnt this widget just to make scrollable widgets that have to few items (and would therefor not be scrollable atm) scrollable.

I am also comparing scrolling in Flutter with native scrolling on iOS and there seem to be lots of missing frames here and there in Flutter and sometimes longer pauses. I am running on an iPhone X so its not some low end device. I heard that ListViewTiles are NOT recycled in Flutter, is this true? and if so, Why ? In game development, where every frame counts it is very important to have object pools to be able to show and hide render-objects fast without destroying and creating objects all the time.

**Comment by lirbank** [comment]
Are you running the app in debug or release mode? On my Moto x4 (kinda low end device) scrolling is jerky in debug mode, but once compiled for production it is just as smooth as Native apps.

```sh
flutter run --release
```

**Comment by kraa** [comment]
Yeah, release build (no debug label top right corner of screen)

**Comment by tirnovali** [comment]
same problem here

**Comment by Eimji** [comment]
I also confirm lag issues on iPhone X

**Comment by nuc134r** [comment]
Confirming lags on Xiaomi Redmi 4X, Android 7.1.2. Release build, latest Flutter version. Is there any way to have native performance while scrolling?

**Comment by totalerex** [comment]
I confirm the issue described. Even more, after hours of investigation of such issue in my app, I'm able to reproduce it in every single Flutter app in the Store.
What I found is that Inkwell and GestureDetector are related to it:
- No Inkwell/GestureDetector - no issue at all
- if there is a widget wrapped with Inkwell or GestureDetector  (even if they are not used) it depends how complex rest of the three. Sometimes if you wrap with additional Inkwell with empty onTap: it may help a little, but doesn't eliminate the problem completely.
For sure I checked my app Widgets for an unexpected calls of "build" method - there is no issue.

**Comment by nuc134r** [comment]
@totalerex 

Flutter has a demo app called Gallery. Sources are in this repo under "examples" folder but they also have it published in Play market. In 'Studies' section they have "Shrine" app demo. After mock login screen you're faced with horizontal (previously vertical) scrolling view with products. It scrolls so smooth that it actually frustrates me. It's that kind of scrolling performance I actually expect but can't achieve. Never thought scrolling can be an issue but it turns out to be.

Can you take a look? It actually uses GestureDetector but still smooth almost like native.

Here is the ListView:
https://github.com/flutter/flutter/blob/master/examples/flutter_gallery/lib/demo/shrine/supplemental/asymmetric_view.dart

Widgets that are placed inside:
https://github.com/flutter/flutter/blob/4857267e6588a0bfad1c537ef0939c26fb4639b6/examples/flutter_gallery/lib/demo/shrine/supplemental/product_columns.dart

And finally inner Widget with the GestureDetector:
https://github.com/flutter/flutter/blob/4857267e6588a0bfad1c537ef0939c26fb4639b6/examples/flutter_gallery/lib/demo/shrine/supplemental/product_card.dart

**Comment by totalerex** [comment]
Gallery was the first app I checked.  Flutter Galery installed from Store has the same issue, in Material and Cupertino screen. It looks like the problem  in Cupertino is not so big but visible also. I have not tried to build and run Gallery, since I can't see any reason why it should behave differently.

**Comment by KlausJokinen** [comment]
Same issue. ListaView builder scrolling is choppy

**Comment by totalerex** [comment]
@nuc134r
Shrine - the same issue. 
if I scroll with some "scroll pattern" slowly then issue is visible.

**Comment by totalerex** [comment]
I did some more investigation. So if we are talking about jerk on 11th second of original GIF, it is exactly the problem I was straggling and I found the solution 

Just to remember - issue is there if ListView items contain InkWell/GestureDetector.

1. If scrolling moved not fast (there are some corner cases as well), tap detection starts and blocks scroll events.
2. After movement more than kTouchSlop = 18 logical pixels (flutter\lib\src\gestures\constants.dart).
is detected, tap detection cancels.
3. Scroll events are unblocked.
4. First Scroll event received by all listeners has offset of kTouchSlop = 18+ logical pixels! 

It is **totally** wrong! Scrolling should start with ~0 offset of first scroll event. It is exactly like native Android scroll behaves it is what my eyes can see.

So my solution to detect first scroll event and set the offset to 0. I did it for internal usage only, for sure, in wrong place from architectural point of view, but it works very smoothly! same way as native scrolling in Android, for sure there is ~18 pixels delay in scrolling in fact, but it looks very nice.

It is the change in flutter\lib\src\widgets\scroll_position_with_single_context.dart 

![image](https://user-images.githubusercontent.com/42652300/52437958-d12f5a00-2b20-11e9-8c03-e991f9444b60.png)

**Comment by dukaric1991** [comment]
Does this last advice improve scroll performance?

**Comment by Toolenaar** [comment]
@dukaric1991 Not for my, tried that and I still get very choppy scrolling. I did some experiments with different ways of building a list And I found the best scroll performance with using a CustomScrollView with a list of Slivers (not a builder). But I think this wont hold up memory wise when the list gets very long.

`
List<Widget> slivers = _snaps.map((f) => SliverToBoxAdapter(child: SnapListItem(),)).toList();
return CustomScrollView(slivers: slivers);
`

**Comment by frasza** [comment]
Hopefully this gets address soon, it's also affecting my experience on iPhone XS using listview builder

**Comment by dukaric1991** [comment]
I found a temporary solution. Try to wrap your listView items/widgets in IgnorePointer widget with parameter ignoring: true. But if you have any "clickable" elements in your listView item/widget, try to wrap only parts of item that are not "clickable" in IgnorePointer, beacause otherwise the click/tap action will not be recognized. 

`
IgnorePointer(
       ignoring: true,
       child: Container(),
)
`

**Comment by ihainan** [comment]
@dukaric1991 😔 Not working for me on iPhone X with the latest Flutter...scrolling is still very choppy and laggy...

**Comment by ishaanbahal** [comment]
Facing this issue quite a bit but in my case the list items are also loading images, so the scroll is extremely choppy. On iOS it is slightly better, but on android it is horrible. Tested on OnePlus 5T (used to be a powerful device till last year) with android 9.0.1.

**Comment by wilin52** [comment]
I have a device Huawei Mate9 Pro ,scrolls not smoothly but it's better in Samsung s8/ OnePlus 3T.
Does it performs differently in various devices?

I create a sample Repository: No InkWell inside list . And using AlwaysScrollableScrollPhysics, but didn't work.
https://github.com/wilin52/custom_scroll_view_demo_scroll_not_smoothly

There's a video in directory screen_record.
https://github.com/wilin52/custom_scroll_view_demo_scroll_not_smoothly/blob/master/screen_record/screen_record.mp4

**Comment by k00na** [comment]
`flutter run --release `
I'm testing on Samsung S8+ and during development the ListView builder is indeed lagging. But I can confirm what @lirbank stated and after running flutter run --release my listview was smooth again.

**Comment by allanwolski** [comment]
We published our app and got this same problem.
Can we expect a quick solution to this issue?

**Comment by wemped** [comment]
This is our team's biggest blocker to using Flutter. Most of the views in our app are based on ListView and we're seeing pretty significant frame skipping and scroll jitters even w/ the most basic layouts. Would love to find a way around these performance issues, or see them slated for a specific milestone soon.

Some of the things we've tried (all while running in Release Mode on a iPhone XS device), all with similarly poor results
- removing images from the ListView items
- removing gesture detectors and inkwells from the ListView items
- using non-builder ListViews w/ a small amount (~10) of items
- using just a Container w/ a static height for all items
- trying out the Flutter Gallery app Shrine and ListView examples

Seems like the issues are less bad when the scrolling is fast, but just slowly dragging the view causes a lot of jitter.

Going back and forth between a Flutter application's ListView and a native TableView/RecyclerView is night and day. 

Here's my `flutter doctor -v` output if it's helpful at all

[✓] Flutter (Channel stable, v1.2.1, on Mac OS X 10.14.4 18E226, locale en-US)
    • Flutter version 1.2.1 at /Users/dwempe/flutter
    • Framework revision 8661d8aecd (8 weeks ago), 2019-02-14 19:19:53 -0800
    • Engine revision 3757390fa4
    • Dart version 2.1.2 (build 2.1.2-dev.0.0 0a7dcf17eb)

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/dwempe/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: /Applications/Android Studio 2.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 10.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.2, Build version 10E125
    • ios-deploy 1.9.4
    • CocoaPods version 1.5.3

[✓] Android Studio (version 3.3)
    • Android Studio at /Applications/Android Studio 2.app/Contents
    • Flutter plugin version 34.0.1
    • Dart plugin version 182.5215
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)

[✓] VS Code (version 1.33.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 2.24.0

[✓] Connected device (2 available)
    • Android SDK built for x86 • emulator-5554             • android-x86 • Android 8.0.0 (API 26)
      (emulator)
    • _________ iPhone      • 00008020-000B25CE0298002E • ios         • iOS 12.2

**Comment by coreysprague** [comment]
@wemped - My team has been struggling with this as well. Anecdotally, the scroll performance is worse on iPhone X & iPhone XS models than it is on iPhone XR or even older phones such as the iPhone 7. Do you observe the same?

One theory our team has is that it might be related to the 120hz gesture sample rate on those particular models. (https://www.macworld.com/article/3235709/iphone-x-samples-touch-input-at-120hz-for-faster-smoother-response.html) -- i.e. maybe the flutter engine is being fed twice as many gesture events on these devices (could explain why slow dragging performs much worse than a fling). We have no factual basis for this theory though and haven't even dug into the code to try to understand if that could be a contributing factor -- it's just a hunch. Would love if someone on the flutter team could chime in.

**Comment by wemped** [comment]
@coreysprague - Thanks for the tip! I dug out my old iPhone 7 and can confirm the scrolling performance is hugely improved.

That's definitely an interesting theory. I imagine this warrants a separate github issue so I made one here https://github.com/flutter/flutter/issues/31086.

If we get some 👍on the new issue we may be more likely to get a response.

**Comment by x4080** [comment]
I wondered why list view fps < native list view since flutter is native compiled (Android is worse than iOS
 counterpart)

**Comment by anasred** [comment]
after running 
`flutter run --release`
then follow the instructions to run it on the simulator or on your real phone


I noticed that the performance increased by 200% 🤩. I have a native app and I'm building a new one using Flutter.

Flutter even faster than the native !!

each element is (in both Flutter and native):
a card with one image, id, and title

**Comment by BlueskyFR** [comment]
Same issue here, has to be fixed ASAP :-/

**Comment by x4080** [comment]
@anasred Do you use iOS or android ? iOS is smooth android not 60fps

**Comment by FullstackJack** [comment]
inKino app looks pretty smooth to me on Google Pixel 2 XL. https://github.com/roughike/inKino you can download it from app store to try. Does anyone see jitter on that app?

**Comment by Hixie** [comment]
This issue described in https://github.com/flutter/flutter/issues/22314#issuecomment-461569961 was fixed recently (should be in the beta channel if you want to try it; may in fact already be in stable, I forget when we fixed it exactly). 

If you still see this today on `master` please file a new bug, including:

- the source code of a demo app that we can use to reproduce this
- precise steps to reproduce it
- the exact device on which we should reproduce it

...and we will try to look into this in more detail. It's hard, from this bug, to tell exactly what we should be looking at since there are so many reports and devices mixed together.

**Comment by nuc134r** [comment]
@Hixie
> This issue described in #22314 (comment)
22314 (comment) quote:
> So if we are talking about jerk on 11th second of original GIF, it is exactly the problem I was struggling and I found the solution

That's native reddit app on 11th second so it's not relevant to flutter scrolling.

**Comment by tudor07** [comment]
I am having the same issue.
On iOS the scroll it's quite good but on Android it is absolutely horrible.
Any updates on this issue @Hixie ?

**Comment by tudor07** [comment]
@totalerex can you give more insight on how you solved the issue ? I don't quite understand from your comment

**Comment by totalerex** [comment]
@tudor07, issue I mentioned, was fixed in framework some time ago (not reproducible now), so @Hixie is absolutely right. 
But as I see now, it was corner case just on start scrolling. Main issue with scrolling jerking on some devices still exists.

**Comment by Gatunox** [comment]
Issue is still present on iOS

**Comment by sssalib42** [comment]
If you are dealing with static data of very low number of list items, you can speed the list up using
```
ListView(
          addAutomaticKeepAlives: true,
          ...
)
```

**Comment by Gatunox** [comment]
ssalib42 thanks for the help, I've tested it and I saw a good improvement, but the problem persists

**Comment by sssalib42** [comment]
True.. you may see even more improvement when running in release mode if you haven’t already...

I’m not an expert on flutter but I like the idea of it very much, I want to see it grow!

My suggestion to the flutter team is to allow more control on the “virtual scrolling”..
For example if there is a way to draw, say the next 5 items, always ahead of the user scrolling, or if my specific list view requires it, up to 10 items pre-drawn, in both scrolling directions, ready to view!

Also, most frameworks reuse widget as oppose to completely destroying them, and redrawing them from scratch.

List view is like the new hello world of functioning apps.. we should make it right, l think.

**Comment by stychu** [comment]
I have the same problem. Listview just simply lags as hell... I've used Card component from official Flutter gallery repo to display Cards as a scrollable list and it lags very hard...

**Comment by stychu** [comment]
Found the problem in my use case. It seems like there is a bug with High res images and I have had made listview with cards containing high res images which were causing stutters and hang the app when scrolling and moving from list screen back and forth. Once I changed images to lower resolution all works great and very fluent now.

**Comment by KlausJokinen** [comment]
@stychu
What was the picture size before and after?

**Comment by Hixie** [comment]
We believe we landed a fix for this a few months ago.

This bug at this point is not actionable for us, because lots of people have commented very vague requests.

I recommend that everyone seeing this each file a new bug with the following:

* A `main.dart` file that clearly shows the problem (that we can drop into a `flutter create` app).
* Clear steps for how to reproduce the problem with that file.
* A clear explanation (ideally with a video) showing the exact issue that you are referring to.
* the output of `flutter doctor -v`.

Thanks.

**Comment by Chamuth** [comment]
> after running
> `flutter run --release`
> then follow the instructions to run it on the simulator or on your real phone
> 
> I noticed that the performance increased by 200% 🤩. I have a native app and I'm building a new one using Flutter.
> 
> Flutter even faster than the native !!
> 
> each element is (in both Flutter and native):
> a card with one image, id, and title

--debug: *slow and lame*
--release: I'm fast as fu** boiii!

**Comment by MihaMarkic** [comment]
@Chamuth Of course debug is slow, that's by design as it runs as interpreter.

**Comment by Gatunox** [comment]
is layout of elements are complex, scroll is laggy even on release mode

**Comment by quantosapplications** [comment]
Hi, any update on this? I tried all the solutions proposed in this post with no success..

**Comment by unclexiao** [comment]
After switching to the `master` branch, this problem seems to be resolved. Will the new version that will be released solve this problem? 

I am very much looking forward to the surprise.

**Comment by KlausJokinen** [comment]
> After switching to the `master` branch, this problem seems to be resolved. Will the new version that will be released solve this problem?
> 
> I am very much looking forward to the surprise.

Do you know any specific commits that fixed this issue?

**Comment by unclexiao** [comment]
@KlausJokinen Yes,  and here here is the [PR](https://github.com/flutter/engine/pull/12280), and the [Commit](https://github.com/flutter/flutter/commit/b9a34dec7eab35abae3d2e0be37a7d73f9d8aceb)

**Comment by muhammadsaddamnur** [comment]
same issue :(

**Comment by fzyzcjy** [comment]
Hi, is there any updates -- does it run as smooth as native or still not well? Thanks!

**Comment by nuc134r** [comment]
> Hi, is there any updates -- does it run as smooth as native or still not well? Thanks!

Still not even on hi end Android devices. Or at least not in all cases.

**Comment by fzyzcjy** [comment]
> > Hi, is there any updates -- does it run as smooth as native or still not well? Thanks!
> 
> Still not even on hi end Android devices. Or at least not in all cases.

Could you please give a minimum reproducible sample so I can compile and see? 

*Why* does this happen?? I think Lists are one of the most fundamental things... How can it be non smooth?

**Comment by unclexiao** [comment]
> Hi, is there any updates -- does it run as smooth as native or still not well? Thanks!
 
Fixed in v1.12.13 , you should  update to newest stable channel  @fzyzcjy

**Comment by iamarnas** [comment]
> > Hi, is there any updates -- does it run as smooth as native or still not well? Thanks!
> 
> Fixed in v1.12.13 , you should update to newest stable channel @fzyzcjy



> > Hi, is there any updates -- does it run as smooth as native or still not well? Thanks!
> 
> Fixed in v1.12.13 , you should update to newest stable channel @fzyzcjy

It's not fixed yet, list scroll smoothly until you not use Expansiontile, when you expand you list and scrolling then scroll indicator shaking if you use it. I tried fix it in all ways but not found good solution. So if you want have smooth list then not use Expansiontile with long expanded list.

**Comment by fzyzcjy** [comment]
@unclexiao @iamarnas Thanks very much!

**Comment by cmpeguerog** [comment]
I test the get started project and the scrolling is bad very bad nothing compared with the native.

**Comment by fzyzcjy** [comment]
@cmpeguerog May I know which get started project, or could you plz share the code so I can try that. Thanks!

**Comment by cmpeguerog** [comment]
@fzyzcjy [Get Started](https://flutter.dev/docs/get-started/codelab) . Every time someone writes that the problem is already solved in a release, I check  that code in the hope that it's true but since I was told about Flutter the error has been there and it's been a while.
Phones: **Kyocera, Motorola E4**

**Comment by fzyzcjy** [comment]
@cmpeguerog Ah let me try it! By the way, do you use *production* mode, or using debug mode which is said to be slower?

**Comment by cmpeguerog** [comment]
**Debug**, but i think that for better user experience should work fine for it too.

**Comment by fzyzcjy** [comment]
@cmpeguerog Is it still non-smooth in release mode? Because flutter said that debug vs release will have a very big difference in their documents

**Comment by cmpeguerog** [comment]
@fzyzcjy I will test it and let u know

**Comment by MihaMarkic** [comment]
Guys, always test performance issues in *release* mode as the code is compiled while in debug it is interpreted. Debug tests don't say anything.

**Comment by deniza** [comment]
I am still experiencing this issue while using latest stable release as of today (1.12.13+hotfix.8). I found that forcing widget tree to repaint itself while scrolling somehow fixes the problem. To quickly test this; please add a dummy animated widget to your widget tree using TweenAnimationBuilder:
```dart
TweenAnimationBuilder(
  duration: Duration(seconds: 10),
  tween: Tween<double>(begin: 0, end: 1000),
  builder: (context, value, widget) => Container(),
)
```
For 10 seconds your list should scroll smoothly. Same issue exits while dragging (ie using Draggable). I am very new to Flutter. So please someone experienced suggest a better way to force repainting widgets while scrolling/dragging.

**Comment by KlausJokinen** [comment]
> I am still experiencing this issue while using latest stable release as of today (1.12.13+hotfix.8). I found that forcing widget tree to repaint itself while scrolling somehow fixes the problem. To quickly test this; please add a dummy animated widget to your widget tree using TweenAnimationBuilder:
> 
> ```dart
> TweenAnimationBuilder(
>   duration: Duration(seconds: 10),
>   tween: Tween<double>(begin: 0, end: 1000),
>   builder: (context, value, widget) => Container(),
> )
> ```
> 
> For 10 seconds your list should scroll smoothly. Same issue exits while dragging (ie using Draggable). I am very new to Flutter. So please someone experienced suggest a better way to force repainting widgets while scrolling/dragging.

Can you retry on master?

**Comment by deniza** [comment]
Switching to master channel (1.13.6-pre.22) does not resolve this issue, at least for me. But repaint hack still works.

**Comment by deniza** [comment]
@KlausJokinen I got same results using "Flutter 1.15.4-pre.124 • channel master •", release build running on an actual device. Scrolling and dragging is "jaggy" on both lower end devices like Samsung Galaxy S3 (GT I9300) and high-end devices like iPhone 11.

**Comment by esDotDev** [comment]
I've added a semi-related issue here with a specific example:
https://github.com/flutter/flutter/issues/52207

In this case, it's related to performance of JumpTo when using many list items, without a defined itemExtent.

**Comment by sirily11** [comment]
When I have my list with 200 more items which all have image within each list row. The list view scroll so slow and finally causes accidental exit

**Comment by ItsPierre** [comment]
> When I have my list with 200 more items which all have image within each list row. The list view scroll so slow and finally causes accidental exit

If I understand this right, you are building the whole list with all 200 items at once? If so, thats not a good idea, you should use the builder to only build the viewed items.

**Comment by sirily11** [comment]
> > When I have my list with 200 more items which all have image within each list row. The list view scroll so slow and finally causes accidental exit
> 
> If I understand this right, you are building the whole list with all 200 items at once? If so, thats not a good idea, you should use the builder to only build the viewed items.

Nope. I used list view builder to build the list. And i tried the solution above but still has the issue.

**Comment by Martibis** [comment]
ListView.builder still seems choppy when you scroll than release finger. As long as finger is on the screen this does not seem to be an issue for me. The chopping starts to occur when the scrolling starts to slow down. I have tried all of the suggestions above and none seem to properly fix this issue for me.

**Comment by cmpeguerog** [comment]
I test, in release mode, the get-started project for infinite scroll and it still laggy

**Comment by Blindjoedeath** [comment]
There is interesting thing. Custom scroll physics extended from BouncingScrollPhysics with overidden tolerance with velocity = 0.0, makes first drag exactly in the position it starts. 

```
class CustomScrollPhysics extends BouncingScrollPhysics {
  const CustomScrollPhysics({ScrollPhysics parent})
      : super(parent: parent);

  @override
  Tolerance get tolerance => Tolerance(velocity: 0.0);

  @override
  CustomScrollPhysics applyTo(ScrollPhysics ancestor) {
    return CustomScrollPhysics(parent: buildParent(ancestor));
  }
}
```


It looks very familiar with native iOS scrolling, but is locks gesture recognitions in list items. So if you need to create scrollable list, without tappable items, you can use it  ¯\\\_(ツ)\_/¯

P.S. If one day this problem would be solved, i will never back in native development, and only use Flutter ❤️

**Comment by fadhlisulaimi** [comment]
In android, if u are using phone that has 90hz/120hz refresh rate, the scrolling will be laggy and it looks obvious

**Comment by dukaric1991** [comment]
I have S20+ with 120hz enabled and its buttery smooth, 120FPS all the time with no lag.

But previously i had S8+ and it was obviously laggy...

**Comment by sirily11** [comment]
Yes. I tested the app in both android and iOS version. iOS will get slower
scrolling when items increases to 300 but android will around 600

Muhammad Adil <notifications@github.com>于2020年4月6日 周一下午3:13写道：

**Comment by firatcetiner** [comment]
This issue is also referred at #22314 and I don't think it is related to the` ScrollBar().`

Apparently setting the ListView's `cacheExtent `property explicitly can be a decent workaround. See the [documentation](https://api.flutter.dev/flutter/widgets/ListView/cacheExtent.html).

In my case, the average height of a child in the list is around 300. So with `cacheExtent `set to 10000.0 the scroll is much more performant.

For calculating the cacheExtent value I'm simply make an equation:
**Inputs:**
Item Count (Amount of items that we want to cache in the viewport) = 10 (This is enough for my list)
Avg. size for a child = 300
Height of current context (`MediaQuery.of(context).size.heigh`t) = 640
Height of device (pixels) = 1920

**Output: cacheExtent**

Then the equation will be as
**_cacheExtent = (Height of device / Height of current context) * (Avg. size for a child) * Item Count_**

**Comment by x4080** [comment]
@firatcetiner Thanks for the information, do you have result with the memory usage with your config ?

**Comment by klaszlo8207** [comment]
Any new on this? Please it is very fustrating

**Comment by pilotviper134** [comment]
The problem still appears on my Huawei P30 Pro device. Using cacheExtent slightly helped with performance. However, the list can still be trimmed at the very end of the scrolling.

**Comment by Renatinaveen** [comment]
https://drive.google.com/file/d/1a3ZJ099XSJ-lvHDc2DVadExAYT3kAM3r/view?usp=drivesdk
That's on iphone 6s

https://drive.google.com/file/d/1w5dEXmKuiKXadzXSxiVbb2GjR05htY6Z/view?usp=drivesdk
This is OnePlus 6t. 

I don't understand what the problem is while scrolling up I get this issue.

**Comment by leszekkrol** [comment]
Why this issue is closed? Problem still occurred

**Comment by iamarnas** [comment]
@leszekkrol It's quick fix. They close many issues for no reason.

**Comment by fadhlisulaimi** [comment]
I tried switching to dev channel and updating to latest version.
Not sure if its just a placebo effort but i feel the scrolling is smoother.

**Comment by firatcetiner** [comment]
> @firatcetiner Thanks for the information, do you have result with the memory usage with your config ?

Not very good actually. Every item in the list consist of:
- A low-resolution image (50x50),
- 5 different `Text` widgets, one of the may text includes a long text (400 characters) sometimes,
- 4 buttons in a row

The structure is like this: Container>Column>[Row[Text, Image], Text, Text, Buttons]

With the cacheExtent = 10000.0 and assuming the average height of a child widget is 600 pixels, it should cache at least 15 items per list. Since I have 4 separate lists on different parts of the app at the same time, this means at least 60 items are in the cache. Assuming the cache size is determined for just only one direction Up or Down but not both.

In this case, the memory usage is around ~350MB. I'm not sure this is okay but it seems like a lot.

**Comment by filiph** [comment]
Hi all, this is an outdated issue, and the explanation to it is unfortunately buried in the noise. Moreover, GitHub collapses the middle of the thread, which [includes](https://github.com/flutter/flutter/issues/22314#issuecomment-527683483) the reason for closing this issue, and it's really easy to overlook it.

![Screen Shot 2020-05-11 at 11 00 38 AM](https://user-images.githubusercontent.com/919717/81595183-e542b680-9376-11ea-9c5e-25eaf5c55974.png)

I want to summarize some things for better discoverability:

* The initial issue has been fixed months ago.
* Please do not look at performance in `debug` mode. It is understood that `debug` mode is much slower than `profile` or `release`. There's only so much the Flutter SDK can do while also allowing for hot reloading, failing fast, etc. If you want to push the Flutter team to make `debug` mode faster in a specific scenario, please file a separate issue.
* If you have a scrolling performance problem, [do file a separate bug](https://github.com/flutter/flutter/issues/new/choose). A video is helpful. A piece of code that reproduces the issue is _extremely_ helpful.
* If you're new to Flutter and wondering about performance, you should know that it is very much possible to have smooth scrolling in Flutter. Many apps out there have complex ListViews and they scroll without missing a frame. On the other hand, it is also possible that Flutter has bugs that make scrolling slow in a particular case (if you suspect this, please file a bug), and it's also possible to make scrolling janky via other code (if you suspect this, please read [flutter.dev/docs/perf/rendering](https://flutter.dev/docs/perf/rendering) or [watch my video](https://www.youtube.com/watch?v=vVg9It7cOfY) from Flutter Europe)

@Hixie, @liyuqian: Would it make sense to rename this issue to something more specific? Would it make sense to close comments on this issue so people don't accidentally bury this information again?

**Comment by esDotDev** [comment]
I have filed a fairly specific issue:
https://github.com/flutter/flutter/issues/52207

It is quite easy to get janky scrolling performance once you attach it to a Scrollbar that calls JumpTo.
1. Make a large list > 5000 items with variable size height rows (ie, don't set itemExtent)
2. Scroll rapidly back and forth

Builder just doesn't seem optimized for successive calls to JumpTo like a Scrollbar does.

**Comment by liyuqian** [comment]
@filiph : I also lost track of what this issue is about and what PR fixed it... Ian's comments https://github.com/flutter/flutter/issues/22314#issuecomment-527683483 and https://github.com/flutter/flutter/issues/22314#issuecomment-488156243 are the closest that I find. @Hixie : I recently noticed "lock bot" sending me a lot of emails. Shall we also lock this issue to incentivize folks to open new issues?

**Comment by leszekkrol** [comment]
Any update in that case?

**Comment by Martibis** [comment]
This is a massive problem a lot of people are facing and key to Flutters adoption in my opinion. You advertise smooth fps, but a basic functionality like scrolling dynamic content can not get the advertised native level smoothness.

**Comment by liyuqian** [comment]
I've changed the title per @filiph 's suggestion to clarify that this issue is about input events handling and it's fixed according to comment ([1]) and ([2]).

If there are other performance issues related to ListView, please create a new issue following the suggestion in comment ([1]) and ([2]).

We understand your performance concerns and we'll definitely work very hard to fix it. It's just difficult for us to follow an old Github issue with a lot of unrelated comments. So I'm locking down this issue to encourage creating new and focused issues with details of how to reproduce the performance issue. Please feel free to reference me by @liyuqian in the new issues.

[1]: https://github.com/flutter/flutter/issues/22314#issuecomment-527683483
[2]: https://github.com/flutter/flutter/issues/22314#issuecomment-488156243
